### PR TITLE
Environment status checks

### PR DESCRIPTION
**Pull request by RazTamir**
- Compare env status before and after test execution for leftovers
 - New installer version 4.1.2

### CODE REVIEW

**Comment by RazTamir** [comment]
A sample output:

            raise exceptions.LeftoversDetected(
>               f"\nThere are leftovers in the environment after test case:"
                f"\nResources added: {leftovers['Leftovers added']}"
                f"\nResources removed: {leftovers['Leftovers removed']}"
            )
E           ocs.exceptions.LeftoversDetected: 
E           There are leftovers in the environment after test case:
E           Resources added: [{'sc': {'new_value': '123456789'}}]
E           Resources removed: []

tests/test_environment_check.py:127: LeftoversDetected

File: utility/environment_check.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    leftovers = {'Leftovers added': [], 'Leftovers removed': []}
+    for kind in diffs_list:
+        if ADDED_RESOURCE in kind[''.join(kind.keys())]:
+            leftovers['Leftovers added'].append({
+                ''.join(kind.keys()):
+                kind[''.join(kind.keys())][ADDED_RESOURCE][
+                    ''.join(kind[''.join(kind.keys())][ADDED_RESOURCE])
+                ]
+            })
+            leftover_detected = True
```
Comment (ebenahar) [inline utility/environment_check.py]: if this one is True, there is no reason to check the removed resources. I suggest checking if there are removed resources only in case there are no added resources

Comment (ebenahar) [reply utility/environment_check.py]: however, you still want to check if there is anything to append to leftovers['Leftovers removed'] even if the previous leftover detection result was true.
So maybe having to leftover_detected args, one for added and one for removed

Comment (RazTamir) [reply utility/environment_check.py]: I'm not sure I follow:
regarding the first comment:
I can create a secret and not delete it (leftover added) and delete a pod from the env by mistake (leftover removed).
Regarding the second comment:
I will print everything in any case (for kind in diffs_list)

Comment (ebenahar) [reply utility/environment_check.py]: @RazTamir See my edit to the first comment

Comment (ebenahar) [reply utility/environment_check.py]: @RazTamir my mistake :) in case there are added resources and no removed ones, leftover_detected will remain True

Comment (RazTamir) [reply utility/environment_check.py]: > if this one is True, there is no reason to check the removed resources. I suggest checking if there are removed resources only in case there are no added resources.
A scenario where I created a resource and forgot to delete it will result in added_resource.
If I also deleted some other resource so I want to print it as well.
I think I'm missing something
> Actually, if there are added leftovers and no removed leftovers, leftover_detected will be False.
Why?
I'm not re-setting levftovers_detected to False ever again - after the first catch of leftover, it will print it
Same, I think I miss something

**Comment by RazTamir** [comment]
More sample output:
I created (manually) PVC and deleted it during the test execution:


E           ocs.exceptions.LeftoversDetected: 
E           There are leftovers in the environment after test case:
E           Resources added: []
E           Resources removed: [{'pvcs': {'apiVersion': 'v1', 'kind': 'PersistentVolumeClaim', 'metadata': {'annotations': {'pv.kubernetes.io/bind-completed': 'yes', 'pv.kubernetes.io/bound-by-controller': 'yes', 'volume.beta.kubernetes.io/storage-provisioner': 'ceph.rook.io/block'}, 'creationTimestamp': '2019-06-15T17:00:03Z', 'finalizers': ['kubernetes.io/pvc-protection'], 'name': 'raz', 'namespace': 'openshift-storage', 'resourceVersion': '360973', 'selfLink': '/api/v1/namespaces/openshift-storage/persistentvolumeclaims/raz', 'uid': '04e28370-8f8f-11e9-9a05-061efedd80e4'}, 'spec': {'accessModes': ['ReadWriteOnce'], 'dataSource': None, 'resources': {'requests': {'storage': '1Gi'}}, 'storageClassName': 'rook-ceph-block', 'volumeMode': 'Filesystem', 'volumeName': 'pvc-04e28370-8f8f-11e9-9a05-061efedd80e4'}, 'status': {'accessModes': ['ReadWriteOnce'], 'capacity': {'storage': '1Gi'}, 'phase': 'Bound'}}}]

**Comment by RazTamir** [comment]
And the opposite:
Started the test execution with clean env and as part of the test, created new PVC:

E           ocs.exceptions.LeftoversDetected: 
E           There are leftovers in the environment after test case:
E           Resources added: [{'pvs': {'apiVersion': 'v1', 'kind': 'PersistentVolume', 'metadata': {'annotations': {'pv.kubernetes.io/provisioned-by': 'ceph.rook.io/block'}, 'creationTimestamp': '2019-06-15T18:02:59Z', 'finalizers': ['kubernetes.io/pv-protection'], 'name': 'pvc-cf6c46b9-8f97-11e9-b80d-061efedd80e4', 'resourceVersion': '381786', 'selfLink': '/api/v1/persistentvolumes/pvc-cf6c46b9-8f97-11e9-b80d-061efedd80e4', 'uid': 'cf9a04e8-8f97-11e9-a7cf-0a4235d7f512'}, 'spec': {'accessModes': ['ReadWriteOnce'], 'capacity': {'storage': '1Gi'}, 'claimRef': {'apiVersion': 'v1', 'kind': 'PersistentVolumeClaim', 'name': 'raz', 'namespace': 'openshift-storage', 'resourceVersion': '381779', 'uid': 'cf6c46b9-8f97-11e9-b80d-061efedd80e4'}, 'flexVolume': {'driver': 'ceph.rook.io/openshift-storage', 'fsType': 'xfs', 'options': {'clusterNamespace': 'openshift-storage', 'dataBlockPool': '', 'image': 'pvc-cf6c46b9-8f97-11e9-b80d-061efedd80e4', 'pool': 'rbd', 'storageClass': 'rook-ceph-block'}}, 'persistentVolumeReclaimPolicy': 'Retain', 'storageClassName': 'rook-ceph-block', 'volumeMode': 'Filesystem'}, 'status': {'phase': 'Bound'}}}, {'pvcs': {'apiVersion': 'v1', 'kind': 'PersistentVolumeClaim', 'metadata': {'annotations': {'pv.kubernetes.io/bind-completed': 'yes', 'pv.kubernetes.io/bound-by-controller': 'yes', 'volume.beta.kubernetes.io/storage-provisioner': 'ceph.rook.io/block'}, 'creationTimestamp': '2019-06-15T18:02:59Z', 'finalizers': ['kubernetes.io/pvc-protection'], 'name': 'raz', 'namespace': 'openshift-storage', 'resourceVersion': '381789', 'selfLink': '/api/v1/namespaces/openshift-storage/persistentvolumeclaims/raz', 'uid': 'cf6c46b9-8f97-11e9-b80d-061efedd80e4'}, 'spec': {'accessModes': ['ReadWriteOnce'], 'dataSource': None, 'resources': {'requests': {'storage': '1Gi'}}, 'storageClassName': 'rook-ceph-block', 'volumeMode': 'Filesystem', 'volumeName': 'pvc-cf6c46b9-8f97-11e9-b80d-061efedd80e4'}, 'status': {'accessModes': ['ReadWriteOnce'], 'capacity': {'storage': '1Gi'}, 'phase': 'Bound'}}}]
E           Resources removed: []

File: utility/environment_check.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+REMOVED_RESOURCE = 'iterable_item_removed'
+
+
+@pytest.fixture(scope='class')
+def environment_checker(request):
+    request.addfinalizer(get_status_after_execution)
+    get_status_before_execution()
+
+
+def get_status_before_execution():
```
Comment (ebenahar) [inline utility/environment_check.py]: Shouldn't we fetch this data out of the CephCluster yaml?
This will ensure we're comparing against the desired state, and will apply also for deploymet - we'll be able to execute this health check after the deployment is done

Comment (RazTamir) [reply utility/environment_check.py]: As discussed f2f, there are a lot of implications by fetching that info from cephcluster yaml.
As we need a quick solution that provides us with what we need, this PR answers that

File: utility/environment_check.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+
+POD = ocp.OCP(kind=constants.POD)
+SC = ocp.OCP(kind=constants.STORAGECLASS)
+CEPHFILESYSTEM = ocp.OCP(kind=constants.CEPHFILESYSTEM)
+CEPHBLOCKPOOL = ocp.OCP(kind=constants.CEPHBLOCKPOOL)
+PV = ocp.OCP(kind=constants.PV)
+PVC = ocp.OCP(kind=constants.PVC)
+SECRET = ocp.OCP(kind=constants.SECRET)
+NS = ocp.OCP(kind=constants.NAMESPACE)
```
Comment (shylesh) [inline utility/environment_check.py]: How about adding deployment, service kinds as well .

File: utility/environment_check.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+"""
```
Comment (shylesh) [inline utility/environment_check.py]: Instead of utility function how about handling it in OCS.create() and OCS.delete() itself. Lets say we add a decorator @register_resource for create() and delete() functions. Every create will register (resource_kind, name, namespace) in a global list and every delete() will remove that element from global list. 

1. If we try to delete the resource which we haven't created, that element will not be in list so raise exception
2. If we created element but forgot to delete, during finalize just scan that list and raise exception.

File: utility/environment_check.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        all_namespaces=True
+    )['items']
+    defaults.ENV_STATUS_PRE['namespace'] = NS.get(all_namespaces=True)['items']
+
+
+def get_status_after_execution():
+    """
+
+    """
+    defaults.ENV_STATUS_POST['pod'] = POD.get(all_namespaces=True)['items']
```
Comment (shylesh) [inline utility/environment_check.py]: Do you want to capture complete details about a resource ? How about simply printing (name, namespace, kind)

File: ocs/ocp.py [hunk 6]
```diff
@@ -88,6 +92,8 @@ def get(self, resource_name='', out_yaml_format=True, selector=None):
             dict: Dictionary represents a returned yaml file
         """
         command = f"get {self.kind} {resource_name}"
+        if all_namespaces and not self.namespace:
+            command += "-A"
```
Comment (shylesh) [inline ocs/ocp.py]: Why all namespaces ? If openshift creates OR deletes its house keeping resources even that would be counted right ? And if user wants to do garbage collection(assuming if user gets leftovers exception and wants to clean up the garbage) I feel it could lead to disaster. Am I missing something ?

File: ocs/exceptions.py [hunk 7]
```diff
@@ -14,12 +14,11 @@ class UnexpectedBehaviour(Exception):
...
 
 
-# TODO: Delete once moved to pytest
-class UnknownStatusOfTestException(Exception):
+class ClassCreationException(Exception):
     pass
 
 
-class ClassCreationException(Exception):
+class LeftoversDetected(Exception):
```
Comment (shylesh) [inline ocs/exceptions.py]: Somehow I am not convinced about this name. I feel following would be more readable.
1. ResourceLeftoversException()
2. GarbageLeftoversException().

Also Leftovers doesn't indicate the resources which deleted by test but not created, because there is nothing leftover there.

**Comment by shylesh** [comment]
I have a small confusion on idea behind this PR.

Are we worried about aws resource consumption after whole test run and hence urge users to cleanup ?
If this is the case then we can make use of labels and annotations, where for any resource we create attach a label or annotations, Something like TO_DELETE=true. At the end of the run just scan all the resources with above annotation or label and delete it.

Please correct me if this totally doesn't make any sense.

File: bin/README.md [hunk 8]
```diff
@@ -1,10 +0,0 @@
-# bin directory
-
-This directory is placed into `PATH` environment variable so that tools located
-here are available to all tests (this is done in
-[`run_ocsci.py`](../run_ocsci.py) wrapper).
-
-The directory is intentionally empty (there should be no other files committed
-there). Command line tools like `oc` and `openshift-install` are installed
-there during deployment (see
-[test_ocs_basic_install.py](../tests/ecosystem/deployment/test_ocs_basic_install.py)).
```
Comment (mbukatov) [inline bin/README.md]: Why is this file deleted? I consider having a README file in every important directory useful to keep it's purpose visible, which helps us with maintenance and usage in the long term.

File: ocs/exceptions.py (hunk 7 above)
Comment (RazTamir) [reply ocs/exceptions.py]: I will change to ResourceLeftoversException().
But what about your "Also" part?

File: ocs/ocp.py (hunk 6 above)
Comment (RazTamir) [reply ocs/ocp.py]: - > Why all namespaces ? If openshift creates OR deletes its house keeping resources even that would be counted right ?
You are right about the first part - I would not expect the OCP will add new entities just like that without any request
 And if user wants to do garbage collection(assuming if user gets leftovers exception and wants to clean up the garbage) I feel it could lead to disaster. Am I missing something ?
So the user should fix the teardown (:

File: utility/environment_check.py (hunk 4 above)
Comment (RazTamir) [reply utility/environment_check.py]: To make it work on every create() delete(), you will need to maintain a map of all resources creation and deletion so you will know what is created as part of the resource.
For example, when I create a cephfilesystem, also 2 mds pods are being created so how can I count it?
I think the approach you are suggesting is much more complicated and doesn't add too much value.

File: utility/environment_check.py (hunk 3 above)
Comment (RazTamir) [reply utility/environment_check.py]: sure, I will add it

File: utility/environment_check.py (hunk 5 above)
Comment (RazTamir) [reply utility/environment_check.py]: yes,

for this initial PR, I'm just checking if new/old resources left/removed after the test finishes but in the future (for now, it is nice to have) we will add more functionality to check also changes in existing resources.
For example, if I'm modifying some existing resource, I want to ensure it is returning to its original values.

File: bin/README.md (hunk 8 above)
Comment (RazTamir) [reply bin/README.md]: I have no idea,
I will restore it in next PR.
Sorry

**Comment by RazTamir** [comment]
> we can make use of labels and annotations, where for any resource we create attach a label or annotations, Something like TO_DELETE=true. At the end of the run just scan all the resources with above annotation or label and delete it.
> 
> Please correct me if this totally doesn't make any sense.

This is an interesting idea.
The only thing that I can think of that might cause issues is extra resources that are being created as part of others (see the previous comment about cephfilesystem)

File: bin/README.md (hunk 8 above)
Comment (mbukatov) [reply bin/README.md]: I added a response here, but can't see it anymore, so sorry if I'm replying twice. I would suggest to just return the file in a separate commit in this pull request, so that when this PR is merged via github's Squash and Merge, the change won't be visible as if nothing happened to the readme file in the final commit.

File: utility/environment_check.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        all_namespaces=True
+    )['items']
+    defaults.ENV_STATUS_POST['pv'] = PV.get(all_namespaces=True)['items']
+    defaults.ENV_STATUS_POST['pvc'] = PVC.get(all_namespaces=True)['items']
+    defaults.ENV_STATUS_POST['secret'] = SECRET.get(
+        all_namespaces=True
+    )['items']
+    defaults.ENV_STATUS_POST['namespace'] = NS.get(
+        all_namespaces=True
+    )['items']
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply utility/environment_check.py]: can you please elaborate on what is the issue?
Not sure I understand this "get all pods for given list of kinds"

File: bin/README.md (hunk 8 above)
Comment (RazTamir) [reply bin/README.md]: I have this file in my local repo and it doesn't seem to be visible here.
In case it will not be part of the merge, I will add it right after the merge

File: utility/environment_check.py (hunk 4 above)
Comment (mbukatov) [reply utility/environment_check.py]: > To make it work on every create() delete(), you will need to maintain a map of all resources creation and deletion so you will know what is created as part of the resource.
> For example, when I create a cephfilesystem, also 2 mds pods are being created so how can I count it?
> I think the approach you are suggesting is much more complicated and doesn't add too much value.

I agree. I would rather keep this simple at first. We can revisit the labeling idea (as noted in different comment) later, if we found it necessary, but right now I don't have enough experience to guess if that will be useful or not.

File: utility/environment_check.py (hunk 1 above)
Comment (mbukatov) [reply utility/environment_check.py]: > if this one is True, there is no reason to check the removed resources. I suggest checking if there are removed resources only in case there are no added resources.
> Actually, if there are added leftovers and no removed leftovers, leftover_detected will be False.

I read it that `leftover_detected` variable covers both addition and removal of something. This could be noted eg. at line:

```
leftover_detected = False
leftovers = {'Leftovers added': [], 'Leftovers removed': []}
```

 And with that in mind, it make sense both cases each time, as these are independent and we care about both.

File: utility/environment_check.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    )
+    namespace_diff = DeepDiff(
+        defaults.ENV_STATUS_PRE['namespace'],
+        defaults.ENV_STATUS_POST['namespace']
+    )
+    diffs_list = [
+        {'pods': pod_diff}, {'sc': sc_diff}, {'cephfs': cephfs_diff},
+        {'cephbp': cephbp_diff}, {'pvs': pv_diff}, {'pvcs': pvc_diff},
+        {'secret': secret_diff}, {'ns': namespace_diff},
+    ]
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply utility/environment_check.py]: Thanks!

I just realized it remained in a debugging form - fixed

**Comment by RazTamir** [comment]
Sample output after new commit:

E           ocs.exceptions.ResourceLeftoversException: 
E           There are leftovers in the environment after test case:
E           Resources added: [{'pvs': {'apiVersion': 'v1', 'kind': 'PersistentVolume', 'metadata': {'annotations': {'pv.kubernetes.io/provisioned-by': 'ceph.rook.io/block'}, 'creationTimestamp': '2019-06-17T12:32:27Z', 'finalizers': ['kubernetes.io/pv-protection'], 'name': 'pvc-f6e62345-90fb-11e9-a630-02ec909f52e0', 'resourceVersion': '1175465', 'selfLink': '/api/v1/persistentvolumes/pvc-f6e62345-90fb-11e9-a630-02ec909f52e0', 'uid': 'f73f17c4-90fb-11e9-a630-02ec909f52e0'}, 'spec': {'accessModes': ['ReadWriteOnce'], 'capacity': {'storage': '1Gi'}, 'claimRef': {'apiVersion': 'v1', 'kind': 'PersistentVolumeClaim', 'name': 'raz', 'namespace': 'openshift-storage', 'resourceVersion': '1174477', 'uid': 'f6e62345-90fb-11e9-a630-02ec909f52e0'}, 'flexVolume': {'driver': 'ceph.rook.io/openshift-storage', 'fsType': 'xfs', 'options': {'clusterNamespace': 'openshift-storage', 'dataBlockPool': '', 'image': 'pvc-f6e62345-90fb-11e9-a630-02ec909f52e0', 'pool': 'rbd', 'storageClass': 'rook-ceph-block'}}, 'persistentVolumeReclaimPolicy': 'Retain', 'storageClassName': 'rook-ceph-block', 'volumeMode': 'Filesystem'}, 'status': {'phase': 'Released'}}}, {'pvcs': {'apiVersion': 'v1', 'kind': 'PersistentVolumeClaim', 'metadata': {'annotations': {'pv.kubernetes.io/bind-completed': 'yes', 'pv.kubernetes.io/bound-by-controller': 'yes', 'volume.beta.kubernetes.io/storage-provisioner': 'ceph.rook.io/block'}, 'creationTimestamp': '2019-06-17T12:39:07Z', 'finalizers': ['kubernetes.io/pvc-protection'], 'name': 'rrr', 'namespace': 'openshift-storage', 'resourceVersion': '1176567', 'selfLink': '/api/v1/namespaces/openshift-storage/persistentvolumeclaims/rrr', 'uid': 'e61bad28-90fc-11e9-a630-02ec909f52e0'}, 'spec': {'accessModes': ['ReadWriteOnce'], 'dataSource': None, 'resources': {'requests': {'storage': '1Gi'}}, 'storageClassName': 'rook-ceph-block', 'volumeMode': 'Filesystem', 'volumeName': 'pvc-e61bad28-90fc-11e9-a630-02ec909f52e0'}, 'status': {'accessModes': ['ReadWriteOnce'], 'capacity': {'storage': '1Gi'}, 'phase': 'Bound'}}}]
E           Resources removed: []

File: ocs/defaults.py [hunk 11]
```diff
@@ -66,6 +66,9 @@
     'bin_dir': './bin',
 }
 
+ENV_STATUS_PRE = {}
+ENV_STATUS_POST = {}
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply ocs/defaults.py]: do you suggest to move it to the exvironment_checker.py file?

File: bin/README.md (hunk 8 above)
Comment (mbukatov) [reply bin/README.md]: It is not going to be part of the merge in it's current state, as the file is removed in commit f92cbaeab758283b2c983d717f991cac7f80cdfa.

If you have the file in your local repository, you don't see it in `git status` output because `bin` directory is in gitignore. I was thinking if there is anything wrong with our code, but didn't noted anything which could have caused this.

Anyway, you can fix it quickly via:

```
$ rm bin/README.md # just in case
$ git checkout c9fc4fd30d8eba7cab3d5787de0eeb6a679a534b -- bin/README.md
$ git commit -m 'reverting README.md removal'
```

Then, when this is merged via "Squash and Merge", the resulting commit won't touch this file as if it were never deleted.

File: ocs/defaults.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbukatov) [reply ocs/defaults.py]: Yes, this would be better as it would prevent the state to be read or changed anywhere in the code. And it would make the code in env. checker self contained.

File: utility/environment_check.py (hunk 4 above)
Comment (shylesh) [reply utility/environment_check.py]: > To make it work on every create() delete(), you will need to maintain a map of all resources creation and deletion so you will know what is created as part of the resource.
> For example, when I create a cephfilesystem, also 2 mds pods are being created so how can I count it?
> I think the approach you are suggesting is much more complicated and doesn't add too much value.

In this specific case it could still work, if you are expecting the test case to create and delete filesystem, we will be tracking the resource like this [(myfs, openshift-storage,CephFilesystem)]. During cleanup just delete myfs so any resource created as part of this fs will be wiped out.

If you want to create fs here and delete it elsewhere then this method won't work.

File: utility/environment_check.py (hunk 5 above)
Comment (shylesh) [reply utility/environment_check.py]: sure

File: ocs/ocp.py (hunk 6 above)
Comment (shylesh) [reply ocs/ocp.py]: Yep, that's why I said house keeping, which may not require any external request although I am not sure about any specific case.

File: utility/environment_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbukatov) [reply utility/environment_check.py]: Method OCP.get() is using inner state of OCP object to get kind and namespace. If there were a function which takes both as an argument instead, we could do something like:

```
kind_list = ["pod", "sc", "cephfs", ... ]
env_status = {}
for kind in kind_list:
    env_status[kind] = ocp_get(kind=kind, all_namespacess=True)
```

without initializing 8 different objects for each kind. It would also make it possible to move the code above into function which could be just called with env_status_pre in setup, and env_status_post in teardown.

File: ocs/defaults.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (RazTamir) [reply ocs/defaults.py]: done

File: utility/environment_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RazTamir) [reply utility/environment_check.py]: creating such a function ocp_get() is pretty easy but I really want this checks to be part of the infrastructure as soon as possible cause there are many existing test cases that not cleaning the env after finish - this will enforce it.

As soon as this will be merged, I will implement such function and change the calls

[the remark this answers is not in the document]
Comment (RazTamir) [reply utility/environment_check.py]: creating such a function ocp_get() is pretty easy but I really want this checks to be part of the infrastructure as soon as possible cause there are many existing test cases that not cleaning the env after finish - this will enforce it.

As soon as this will be merged, I will implement such function and change the calls

**Comment by RazTamir** [comment]
Hi @red-hat-storage/top-level-reviewers ,
I know there are a lot of things to improve in this mechanism and I'll be happy to do so but this PR is pretty urgent as we have many test case PRs coming in that do should not pass the quality level for test case - leaving leftovers in the environment and more bad, deleting existing ones.
So in case there is nothing really important to fix, I'll be happy to merge this PR and all improvements will be done in next PR (feel free to open issues about it and assign it to me)

File: ocsci/testlib.py [hunk 12]
```diff
@@ -1,6 +1,10 @@
 from ocsci.pytest_customization.marks import *  # noqa: F403
+from utility import environment_check as ec
 
 
+@pytest.mark.usefixtures(  # noqa: F405
+    ec.environment_checker.__name__,
```
Comment (andrewschoen) [inline ocsci/testlib.py]: You should only have to use the string 'environment_checker' here. Loading this fixture into ``tests/conftest.py`` makes is automatically accessible.

Also, why wouldn't we want to define this as an ``autouse`` fixture and just have it clean up after each test class has finished executing?

File: utility/environment_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbukatov) [reply utility/environment_check.py]: Agreed, as I noted in the [1st comment here](https://github.com/red-hat-storage/ocs-ci/pull/237/files#r294251297), we don't need to address it in this PR.

File: ocsci/testlib.py (hunk 12 above)
Comment (mbukatov) [reply ocsci/testlib.py]: I agree, the fixture usage is not optimal and what you are suggesting it a canonical pytest way. But since this is not fully followed in code which is already merged, I would not block this PR on this right now. It would be better to agree on change how to use fixtures, change it in all places and then start enforcing it during reviews.

File: ocs/defaults.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbukatov) [reply ocs/defaults.py]: I don't see any change.

[the remark this answers is not in the document]
Comment (RazTamir) [reply ocs/defaults.py]: Pushed now

**Comment by RazTamir** [comment]
So all, please assign all issues to me

File: utility/environment_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (vasukulkarni) [reply utility/environment_check.py]: @mbukatov  It can be just a util function, dont know what it has to do with OCP class design.

**Comment by vasukulkarni** [comment]
@RazTamir I am assuming some testcases are not cleaning up the objects they created which ideally should be testcases cleanup, but as such having this helper is important but I would like it to be configurable too, I will raise an issue for you and will not block this PR so that we can have clean runs.

**Comment by shylesh** [comment]
> > we can make use of labels and annotations, where for any resource we create attach a label or annotations, Something like TO_DELETE=true. At the end of the run just scan all the resources with above annotation or label and delete it.
> > Please correct me if this totally doesn't make any sense.
> 
> This is an interesting idea.
> The only thing that I can think of that might cause issues is extra resources that are being created as part of others (see the previous comment about cephfilesystem)

According to my understanding, consider CephFS creation, it creates few pods and other resources.
If you delete cephfs lets say using its name ocs.delete('myfs') all the resources like pods etc created through this fs resource will be destroyed.
So effectively we need to keep only cephfs reference rather than mds pods created. I don't have complete working knowledge of openshift resources and how it will be used through out tests but I felt this is feasible.

**Comment by RazTamir** [comment]
The "resources will be destroyed" part is something we as QE need to ensure and can't rely on.
Everything else is correct and I think this can be a great approach for teardowns - you can maintain a list of resources you created and just delete as part of your teardown. The env check will ensure everything is ok. 
Please open an issue and assign it to me. 

Thanks

File: ocsci/testlib.py (hunk 12 above)
Comment (mbukatov) [reply ocsci/testlib.py]: This is covered in https://github.com/red-hat-storage/ocs-ci/issues/200

File: utility/environment_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbukatov) [reply utility/environment_check.py]: https://github.com/red-hat-storage/ocs-ci/issues/248

File: utility/environment_check.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+Util for environment check befora and after test to compare and find stale
+leftovers
+"""
+import logging
+import pytest
+
+from ocs import ocp, constants, exceptions
+from deepdiff import DeepDiff
+
+log = logging.getLogger(__name__)
```
Comment (petr-balogh) [inline utility/environment_check.py]: @RazTamir I see that log logger is not used anywhere, didn't want you print some info to the log?

**Comment by RazTamir** [comment]
Only in the raise exception.
In case all good, I want to leave less logs a possible
